rfc/0011-audit-checks.md

```diff
@@ -0,0 +1,354 @@
+
+# Detailed design
+
+The audit check will run at specified intervals and will produce audit reports using the `PolicyReport` format.
+There is going to be one `PolicyReport` object per namespace, plus one `ClusterPolicyReport` that will hold information
+about the cluster-wide resources.
+
+Details about the `PolicyReport` format can be found inside of this [RFC](https://github.com/kubewarden/rfc/pull/13/files).
+
+## Reducing the scope to keep things simple
+
+In the beginning we should reduce the scope of the audit checks to be able to
+ship on time and iterate over the implementation by using real world data.
+
+### No audit for policies targeting `*`
+
+Certain policies can target any kind of Kubernetes resource. The safe
+annotations policy is an example of that.
+
+Targeting all kind of resources would make more complicated to find all the
+Kubernetes resources contained inside of a Namespace that have to be audited.
+
+To simplify things, we can state that policies targeting `*` are not going to be
+evaluated by the audit checks. Users have to be explicit about the
+Kubernetes resources a policy targets.
+
+### No live update of `PolicyResult` objects
+
+The proposed workflow requires that only one process is actively auditing the
+resources of a given Namespace.
+
+That's because:
+
+* The `PolicyResult` of the Namespace is used to reduce the
+  number of auditing evaluations performed
+* Once all the policies and involved resources are audited, the
+  `PolicyReport` audit is going to be overwritten with the results of the audit
+
+Users can still get live data about the rejections/approvals by looking both
+at Prometheus data and Open Telemetry traces.
+
+We should focus on making the process of creating a `PolicyReport` as fast as
+possible. The reports are recreated on a regular basis. If a user wants a fresh
+`PolicyReport` without having to wait for the next scheduled run, we can
+provide a way for the user to trigger this operation.
+
+Finally, if we really have to implement live update of `PolicyReport`, we can
+come up with an alternative architecture that will allow for that. For example,
+we could have all the write operations of `PolicyReport` to go through a
+dedicated "writer" service.
+
+## Creating a `PolicyReport` object
+
+The following section describes how, given a Namespace, we will produce a `PolicyReport` describing the compliance status.
+
+The process to create the `PolicyReport` for a single Namespace can be applied also to generate the `ClusterPolicyReport`, which is the one targeting only cluster-wide resources.
+
+### Step 1: create association between `ClusterAdmissionPolicy` and namespaces
+
+`ClusterAdmissionPolicy` by default inspect events happening inside of all the namespaces, however users can limit their scope
+by using the `spec.namespaceSelector` field.
+
+We need to create a dictionary that holds the following information:
```

Comment:
I would create a dictionary that holds the following information:
* key: namespace
* value: list of `ClusterAdmissionPolicies` relevant to this namespace
This would simplify step 3, because we can easily know all the `ClusterAdmissionPolicies` relevant to a namespace just by getting all values using the namespace as key.